src/agent_workload.py

```diff
@@ -0,0 +1,185 @@
+        # When status is updated, it is likely desirable to have some kind of side effect.
+        self._side_effect()
+
+    def _side_effect(self):
+        logger.debug("Status updated to: %s", self._combined())
+        self._callback(self._combined())
+
+    def _combined(self) -> StatusBase:
+        # Currently there's only one status component, so the combined status is just it.
+        return self.config
+
+
+class WorkloadManager(Object):
+    """Workload manager for grafana agent."""
+
+    CONFIG_PATH = "/etc/grafana-agent.yaml"
+
+    def __init__(
+        self,
+        charm,
+        *,
+        container_name: str,
+        config_getter: Callable[[], Any],
+        status_changed_callback: Callable[[StatusBase], None],
+    ):
+        # Must inherit from ops 'Object' to be able to register events.
+        super().__init__(charm, f"{self.__class__.__name__}-{container_name}")
+
+        # Property to facilitate centralized status update
+        self.status = Status(callback=status_changed_callback)  # pyright: ignore
+
+        self._unit = charm.unit
+
+        self._service_name = self._container_name = container_name
+        self._container = charm.unit.get_container(container_name)
+
+        self._render_config = config_getter
+
+        # turn the container name to a valid Python identifier
+        snake_case_container_name = self._container_name.replace("-", "_")
+        charm.framework.observe(
+            getattr(charm.on, "{}_pebble_ready".format(snake_case_container_name)),
+            self._on_pebble_ready,
+        )
+
+    def _cli_args(self) -> str:
+        """Return the cli arguments to pass to agent.
+
+        Returns:
+            The arguments as a string
+        """
+        return f"-config.file={self.CONFIG_PATH}"
+
+    def _on_pebble_ready(self, _):
+        self._update_config()
+
+        pebble_layer = {
+            "summary": "agent layer",
+            "description": "pebble config layer for Grafana Agent",
+            "services": {
+                "agent": {
+                    "override": "replace",
+                    "summary": "agent",
+                    "command": f"/bin/agent {self._cli_args()}",
+                    "startup": "enabled",
+                },
+            },
+        }
+        self._container.add_layer(self._service_name, pebble_layer, combine=True)
+        self._container.autostart()
+
+        if version := self.version:
+            self._unit.set_workload_version(version)
+        else:
+            logger.debug(
+                "Cannot set workload version at this time: could not get grafana-agent version."
+            )
+
+    def is_ready(self):
+        """Is ready."""
+        return self._container.can_connect()
+
+    @property
+    def version(self) -> Optional[str]:
+        """Returns the version of the agent.
+
+        Returns:
+            A string equal to the agent version
+        """
+        if not self.is_ready:
```

Comment:
`is_ready` is not a property (but imho it should be) so this will always be trivially true.